### PR TITLE
WIP: Add surface abstraction.

### LINKED ISSUE #44: Scope of client-toolkit
What is the scope of client-toolkit? It seems to me that some of the stuff that is handled in winit could be moved over here, like utilities for hidpi handling for example.

I could just use winit of course but I'm interested in experimenting with a new wayland protocol for creating menus, like for example in macosx where they are part of the window manager. In particular I want to create a spacemacs style compositor.

### PR DESCRIPTION
**Pull request by dvc94ch**
Still need to do some more testing

Closes #44

### CODE REVIEW

**Comment by dvc94ch** [comment]
Updated the PR. I think it's much better now and also nicer than the winit implementation.

File: src/surface.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+//! Utility functions for creating dpi aware wayland surfaces.
+
+use std::sync::{Arc, Mutex};
+use wayland_client::Proxy;
+use wayland_client::protocol::wl_surface;
+use wayland_client::protocol::wl_compositor::{RequestsTrait as CompositorRequest};
+use crate::env::Environment;
+
+/// Creates a WlSurface from an Environment.
+/// Takes a callback for notification of dpi changes.
```
Comment (vberger) [inline src/surface.rs]: Please add a newline between the title of the description and its body. Also please expand the documentation to detail how the dpi value is computed (taking the max of the dpi factors of all outputs this surface is displayed on).

File: src/surface.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+pub fn create_surface<F, T>(
+    environment: &Environment,
+    dpi_change: F,
+) -> Proxy<wl_surface::WlSurface>
+where
+    F: Fn(i32) -> T + Send + 'static
```
Comment (vberger) [inline src/surface.rs]: I think this could be an `FnMut` rather than a plain `Fn`, couldn't it?

File: src/surface.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                }
+                dpi_change(scale_factor);
+            }
+        }, Mutex::new(1))
+    }).unwrap()
+}
+
+/// Returns the current dpi factor of a surface.
+pub fn get_dpi_factor(surface: &Proxy<wl_surface::WlSurface>) -> i32 {
+    *surface.user_data::<Mutex<i32>>().unwrap().lock().unwrap()
```
Comment (vberger) [inline src/surface.rs]: We could handle more gracefully the case where the user tries to use this method on a surface that was not created by this helper, in which case the `user_data()` call will return `None`. A possibility would be to replace the first `unwrap()` with an `expect(..)` with a descriptive error message.

File: src/surface.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
+//! Utility functions for creating dpi aware wayland surfaces.
+
+use std::sync::{Arc, Mutex};
+use wayland_client::Proxy;
+use wayland_client::protocol::wl_surface;
+use wayland_client::protocol::wl_compositor::{RequestsTrait as CompositorRequest};
+use crate::env::Environment;
```
Comment (vberger) [inline src/surface.rs]: SCTK tries to remain compatible with older versions of rust, as such we can't use features like `crate::`

**Comment by vberger** [comment]
Perfect! I think this is just missing an entry in the changelog before merging.

**Comment by vberger** [comment]
Looks like there are still some small issues, the travis build is still legitimately red

**Comment by vberger** [comment]
Perfect! Thanks a lot for taking the time to implement this! :+1:

**Comment by vberger** [comment]
With some refactoring (putting the outputs list in the user data) it would be pretty easy to expose a freestanding function to retrieve the list of monitors. Something like

```
fn get_outputs(&Proxy<WlSurface>) -> Vec<Proxy<WlOutput>> { ... }
```

And it' be reasonable to change the closure to pass the surface as argument too.

**Pull request merged**